File: leetcode/1624-largest-substring-between-two-equal-characters/maxLengthBetweenEqualCharacters.go
package leetcode

func maxLengthBetweenEqualCharacters(s string) int {
	var letterFirstIndexMap = make(map[rune]int)
	var length = -1

	for index, letter := range s {
		if firstIndex, ok := letterFirstIndexMap[letter]; ok {
			length = max(length, index-firstIndex-1)
			continue
		}

		letterFirstIndexMap[letter] = index
	}

	return length
}

func max(first int, current int) int {
	if first > current {
		return first
	}
	return current
}

File: leetcode/1624-largest-substring-between-two-equal-characters/maxLengthBetweenEqualCharacters.ts
function maxLengthBetweenEqualCharacters(s: string): number {
  const letterFirstIndexMap = new Map<string, number>();
  let length = -1;

  for (let index = 0; index < s.length; index++) {
    const letter = s[index]!;
    if (!letterFirstIndexMap.has(letter)) {
      letterFirstIndexMap.set(letter, index);
      continue;
    }

    length = Math.max(length, index - letterFirstIndexMap.get(letter)! - 1);
  }

  return length;
}

export default maxLengthBetweenEqualCharacters;
